langchain/output_parsers/json.py

```diff
@@ -2,60 +2,61 @@
 
 import json
 import re
-from typing import List
+from typing import List, Union
 
 from langchain.schema import OutputParserException
 
 
-def parse_json_markdown(json_string: str) -> dict:
+def parse_json_markdown(json_string: str) -> Union[dict, List[dict]]:
     """
-    Parse a JSON string from a Markdown string.
+    Parse JSON objects from a Markdown string.
 
     Args:
         json_string: The Markdown string.
 
     Returns:
-        The parsed JSON object as a Python dictionary.
+        The parsed JSON objects as a Python dictionary or a list of dictionaries.
     """
-    # Try to find JSON string within triple backticks
-    match = re.search(r"```(json)?(.*?)```", json_string, re.DOTALL)
+    # Find all JSON strings within triple backticks
+    matches = re.findall(r"```json\s*({.*?})\s*```", json_string, re.DOTALL)
 
-    # If no match found, assume the entire string is a JSON string
-    if match is None:
-        json_str = json_string
-    else:
-        # If match found, use the content within the backticks
-        json_str = match.group(2)
+    # Parse each JSON string into a Python dictionary
+    parsed_objects = [json.loads(match) for match in matches]
 
-    # Strip whitespace and newlines from the start and end
-    json_str = json_str.strip()
+    # If there's only one JSON object, return it as a dictionary; else, return a list of dictionaries
+    return parsed_objects[0] if len(parsed_objects) == 1 else parsed_objects
 
-    # Parse the JSON string into a Python dictionary
-    parsed = json.loads(json_str)
 
-    return parsed
-
-
-def parse_and_check_json_markdown(text: str, expected_keys: List[str]) -> dict:
+def parse_and_check_json_markdown(
+    text: str, expected_keys: List[str]
+) -> Union[dict, List[dict]]:
     """
-    Parse a JSON string from a Markdown string and check that it
-    contains the expected keys.
+    Parse JSON objects from a Markdown string and check that they
+    contain the expected keys.
 
     Args:
         text: The Markdown string.
-        expected_keys: The expected keys in the JSON string.
+        expected_keys: The expected keys in the JSON objects.
 
     Returns:
-        The parsed JSON object as a Python dictionary.
+        The parsed JSON object(s) as a Python dictionary or a list of dictionaries.
     """
     try:
-        json_obj = parse_json_markdown(text)
+        json_objects = parse_json_markdown(text)
     except json.JSONDecodeError as e:
         raise OutputParserException(f"Got invalid JSON object. Error: {e}")
-    for key in expected_keys:
-        if key not in json_obj:
-            raise OutputParserException(
-                f"Got invalid return object. Expected key `{key}` "
-                f"to be present, but got {json_obj}"
-            )
-    return json_obj
+
+    # Ensure json_objects is always a list
+    if not isinstance(json_objects, list):
```

Comment:
rather than convert to list, and then convert back to list at the end, clearer imo to do

```
if isinstance(json_object, list):
      for obj in json_object:
          _validate_obj(obj)
else:
      _validate_obj(json_obj)
return json_obj
```